src/provider.rs

```diff
@@ -34,30 +40,49 @@ use crate::blobs::{Blob, Collection};
 use crate::protocol::{
     read_lp, write_lp, AuthToken, Closed, Handshake, Request, Res, Response, VERSION,
 };
+use crate::rpc_protocol::{
+    ListRequest, ListResponse, ProvideRequest, ProvideResponse, ProviderRequest, ProviderService,
+    VersionRequest, VersionResponse, WatchRequest, WatchResponse,
+};
 use crate::tls::{self, Keypair, PeerId};
 use crate::util::{self, Hash};
 
 const MAX_CONNECTIONS: u32 = 1024;
 const MAX_STREAMS: u64 = 10;
+const HEALTH_POLL_WAIT: Duration = Duration::from_secs(1);
 
 /// Database containing content-addressed data (blobs or collections).
-#[derive(Debug, Clone)]
-pub struct Database(Arc<HashMap<Hash, BlobOrCollection>>);
+#[derive(Debug, Clone, Default)]
+pub struct Database(Arc<Mutex<HashMap<Hash, BlobOrCollection>>>);
```

Comment:
I'm having a deja-vu.  I'm sure I once wondered if this could be an `RwLock` but I can't remember what the conclusion of that conversation was.

Reply:
RwLock would be fine. I just did the simplest thing possible here. Even used the stdlib mutex instead of my usual parking_lot.

Reply:
Upgraded this to a RwLock